src/main/kotlin/net/maizegenetics/phgv2/pathing/BuildKmerIndex.kt

```diff
@@ -244,96 +244,150 @@ class BuildKmerIndex: CliktCommand(help="Create a kmer index for a HaplotypeGrap
-                }
+                //Do extract out the kmers for the refRange and export to file
+                extractKmersAndExportIndexForRefRange(refRangeToKmerSetMap, refrange, kmerMapToHapIds, refRangeToHapIndexMap, myWriter)
+            }
+        }
 
-                val numberOfHaplotypesInRange = refRangeToHapIndexMap[refrange]?.size?:throw IllegalStateException("Error no haplotypes associated with this refRange.")
-                val numberOfHapSets = hapidKmerHashMap.size
-                if (numberOfHapSets == 0) continue  //skip to next refrange because there are no kmers here
-                val numberOfBitsNeeded = numberOfHaplotypesInRange * numberOfHapSets
-                //encodedHapSets is a BitSet containing all of the Hapsets seen in this reference range
-                //each hapset is encoded with a series of bits equal to the number of haplotypes in the range
-                //with the bits in this hapset set
-                //the offset for each kmer is the offset for its associated hapset
-                val encodedHapSets = BitSet(numberOfBitsNeeded)
-                //make pairs of kmerHash, offset
-                val kmerHashOffsets = mutableListOf<Pair<Long,Int>>()
-
-                var offset = 0
-                val hapidIndex = refRangeToHapIndexMap[refrange]
-                check(hapidIndex != null) {"hapidIndex null for $refrange"}
-
-                hapidKmerHashMap.entries.forEachIndexed { index, entry ->
-                    //this line encodes a haplotype set (hapset)
-                    for (hapid in entry.key) encodedHapSets.set(offset + hapidIndex[hapid]!!)
-                    //this line stores a pair of kmerHash, offset for each kmer mapping to this haplotype set
-                    for (kmerHash in entry.value) kmerHashOffsets.add(Pair(kmerHash,offset))
-                    offset += numberOfHaplotypesInRange
-                }
+        myLogger.info("Saved kmer mapping to $filePath, elapsed time ${(System.nanoTime() - startTime)/1e9} sec")
+    }
 
-                //write the results of the range to a file
-                //line 1: >rangeid
-                myWriter.write(">$refrange\n")
-                //line 2: long1,long2,...,longn (range has hapid sets encoded into a bitSet, which can be stored as longs)
-                myWriter.write("${encodedHapSets.toLongArray().joinToString(",")}\n")
-                //line 3: hash1,offset1,hash2,offset2,...,hashn,offsetn
-                myWriter.write("${kmerHashOffsets.map { "${it.first}@${it.second}" }.joinToString(",")}\n")
+    fun extractKmersAndExportIndexForRefRange(
+        refRangeToKmerSetMap: Map<ReferenceRange, Set<Long>>,
+        refrange: ReferenceRange,
+        kmerMapToHapIds: Long2ObjectOpenHashMap<Set<String>>,
+        refRangeToHapIndexMap: Map<ReferenceRange, Map<String, Int>>,
+        myWriter: BufferedWriter
+    ) {
+        //Check to make sure the refRange maps have the correct keys.
```

Comment:
It may be possible to have a refrange with no kmers. For example if you get a small intergenic refrange with <32 bp or a very short that has no unique kmers. So, maybe a warning rather than an error??

Reply:
I can add this in.